velox/functions/prestosql/StringFunctions.h

```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
+  /// or create a new one if it doesn't exist. Return NULL if the
+  /// specified lang is not supported.
+  static Stemmer* getStemmer(const char* lang) {
+    thread_local std::map<std::string, Stemmer*> stemmers;
+    if (auto found = stemmers.find(lang); found != stemmers.end()) {
+      return found->second;
+    }
+    // Only support ASCII and UTF-8
+    if (auto sbstemmer = sb_stemmer_new(lang, "UTF_8")) {
+      auto rev = new Stemmer(sbstemmer);
+      stemmers[lang] = rev;
+      return rev;
+    }
+    return NULL;
+  }
+
+  /// Get the word stem or NULL if out of memory
+  const char* stem(const std::string& input) {
+    return (const char*)(sb_stemmer_stem(
+        sbstemmer,
+        reinterpret_cast<unsigned char const*>(input.c_str()),
+        input.length()));
+  }
+};
+
+/// word_stem function
+/// word_stem(word) -> varchar
+///     return the stem of the word in the English language
+/// word_stem(word, lang) -> varchar
+///     return the stem of the word in the specificed language
+///
+/// It uses the snowball stemmer library to calculate the stem.
+/// https://snowballstem.org
+/// It provides Java implementation which is used in Presto as well
+/// as C implementation. Therefore, both Presto and Prestimissio
+/// would have the same word stem results.
+template <typename T>
+struct WordStemFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input) {
+    return doCall<false>(result, input);
+  }
+
+  FOLLY_ALWAYS_INLINE void callAscii(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input) {
+    return doCall<true>(result, input);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& lang) {
+    return doCall<false>(result, input, lang.data());
+  }
+
+  FOLLY_ALWAYS_INLINE void callAscii(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& lang) {
+    return doCall<true>(result, input, lang.data());
+  }
+
+  template <bool isAscii>
+  FOLLY_ALWAYS_INLINE void doCall(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const char* lang = "en") {
+    auto stemmer = Stemmer::getStemmer(lang);
+    if (!stemmer) {
+      // language is not supported
+      throw std::invalid_argument(
+          "Unknown stemmer language: \"" + std::string(lang) + "\"");
+    }
+
+    std::string lowerOutput;
+    stringImpl::lower<isAscii>(lowerOutput, input);
+    auto rev = stemmer->stem(lowerOutput);
+    if (rev == NULL) {
+      throw std::runtime_error("out of memory");
```

Comment:
Would you explain how memory allocations are happening here? We need to make sure any large allocations go through Velox's Memory Pool. How can we run out of memory here?

Reply:
enclosed its API doc here:

```
/** Stem a word.
 *
 *  The return value is owned by the stemmer - it must not be freed or
 *  modified, and it will become invalid when the stemmer is called again,
 *  or if the stemmer is freed.
 *
 *  The length of the return value can be obtained using sb_stemmer_length().
 *
 *  If an out-of-memory error occurs, this will return NULL.
 */
const sb_symbol *   sb_stemmer_stem(struct sb_stemmer * stemmer,
                                    const sb_symbol * word, int size);
```

that's why I throw an `out of memory` exception when the returned value is NULL.

Reply:
Let's use VELOX_CHECK_NOT_NULL(stem, "Stemmer library returned a NULL (out-of-memory)."

We need to make it clear that the error comes from Stemmer library and not from Velox core.

Let's also figure out how much memory Stemmer instances use.

Reply:
> Let's use VELOX_CHECK_NOT_NULL(stem, "Stemmer library returned a NULL (out-of-memory)."

Sure thing!

> Let's also figure out how much memory Stemmer instances use.

I guess this would be a tricky one. I will dig more on this.

Reply:
You can write a simple program that uses stemmer lib to stem some words and capture process-wide memory usage before creating stemmer instance, after, then again after processing some words, then after deleting stemmer instance.

https://stackoverflow.com/questions/63166/how-to-determine-cpu-and-memory-consumption-from-inside-a-process

Reply:
I can also look into its implementation and see how and how much it allocates for storing the stem. since each stemmer reuses the memory block, it shouldn't consume that much.